test/grpc/server/router/template_test.exs

```diff
@@ -0,0 +1,175 @@
+               {:"}", []}
+             ] == Template.tokenize("/v1/messages/{message_id}")
+    end
+
+    test "can tokenize variable assignments in bindings" do
+      assert [
+               {:/, []},
+               {:identifier, "v1", []},
+               {:/, []},
+               {:"{", []},
+               {:identifier, "name", []},
+               {:=, []},
+               {:identifier, "messages", []},
+               {:"}", []}
+             ] == Template.tokenize("/v1/{name=messages}")
+    end
+
+    test "can tokenize variable sub-paths in bindings" do
+      assert [
+               {:/, []},
+               {:identifier, "v1", []},
+               {:/, []},
+               {:"{", []},
+               {:identifier, "name", []},
+               {:=, []},
+               {:identifier, "messages", []},
+               {:/, []},
+               {:*, []},
+               {:"}", []}
+             ] == Template.tokenize("/v1/{name=messages/*}")
+    end
+
+    test "can tokenize field paths in bindings" do
+      assert [
+               {:/, []},
+               {:identifier, "v1", []},
+               {:/, []},
+               {:identifier, "messages", []},
+               {:/, []},
+               {:"{", []},
+               {:identifier, "message_id", []},
+               {:"}", []},
+               {:/, []},
+               {:"{", []},
+               {:identifier, "sub.subfield", []},
+               {:"}", []}
+             ] == Template.tokenize("/v1/messages/{message_id}/{sub.subfield}")
+    end
+
+    test "can tokenize single wildcard" do
+      assert [{:/, []}, {:*, []}] == Template.tokenize("/*")
+    end
+
+    test "can tokenize multiple wildcards" do
+      assert [
+               {:/, []},
+               {:*, []},
+               {:/, []},
+               {:*, []},
+               {:*, []}
+             ] == Template.tokenize("/*/**")
+    end
+  end
+
+  describe "parse/3" do
+    test "can parse simple paths" do
+      assert [] ==
+               "/"
+               |> Template.tokenize()
+               |> Template.parse([])
+    end
+
+    test "can parse paths with identifiers" do
+      assert ["v1", "messages"] ==
+               "/v1/messages"
+               |> Template.tokenize()
+               |> Template.parse([])
+    end
+
+    test "can parse paths with 'any'" do
+      assert ["v1", "messages", {:_, []}] ==
+               "/v1/messages/*"
+               |> Template.tokenize()
+               |> Template.parse([])
+    end
+
+    test "can parse paths with 'catch all'" do
+      assert ["v1", "messages", {:__, []}] ==
+               "/v1/messages/**"
+               |> Template.tokenize()
+               |> Template.parse([])
+    end
+
+    test "can parse simple bindings with variables" do
+      assert ["v1", "messages", {:message_id, []}] ==
+               "/v1/messages/{message_id}"
+               |> Template.tokenize()
+               |> Template.parse([])
+    end
```

Comment:
```suggestion
    test "can parse simple bindings with variables" do
      assert ["v1", "messages", {:message_id, []}] ==
               "/v1/messages/{message_id}"
               |> Template.tokenize()
               |> Template.parse([])
    end
    
    test "doesn't crash the beam with malicious input" do
      inputs = 0..:erlang.system_info(:atom_limit) |> Enum.map(&"/v1/messages/{var#{&1}}")
      
      for input <- inputs do
        input |> Template.tokenize() |> Template.parse()
      end
    end
```

this shows that the `String.to_atom` contained in the Template module is an attack vector

Reply:
Are these templates to passed in compile time or are they arbitrarily received through requests?

Reply:
Compile time